core/src/SYCL/Kokkos_SYCL_ParallelFor_Range.hpp

```diff
@@ -17,16 +17,27 @@ template <typename FunctorWrapper, typename Policy>
 struct FunctorWrapperRangePolicyParallelFor {
   using WorkTag = typename Policy::work_tag;
 
+  // We never launch with more than INT_MAX work items which means work items
+  // might execute the functor for multiple indices.
+  // Choosing INT_MAX aligns well with -fsycl-id-queries-fit-in-int.
   void operator()(sycl::item<1> item) const {
-    const typename Policy::index_type id = item.get_linear_id() + m_begin;
-    if constexpr (std::is_void_v<WorkTag>)
-      m_functor_wrapper.get_functor()(id);
-    else
-      m_functor_wrapper.get_functor()(WorkTag(), id);
+    typename Policy::index_type id        = item.get_linear_id() + m_begin;
+    const typename Policy::index_type end = m_work_size + m_begin;
+    while (true) {
+      if constexpr (std::is_void_v<WorkTag>)
+        m_functor_wrapper.get_functor()(id);
+      else
+        m_functor_wrapper.get_functor()(WorkTag(), id);
+      // We need to execute for another index if id + INT_MAX < end, and need
+      // to take care for this check to not overflow.
+      if (end <= INT_MAX || (id >= (end - INT_MAX))) break;
+      id += INT_MAX;
```

Comment:
Not blocking on this but note that we have a `impl/Kokkos_CheckedIntegerOps.hpp` and that it probably make sense to introduce a `add_overflow()` and use it here.
We could do it as a follow up, or as a preliminary PR (not in this PR)

Reply:
Fine